src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs

```diff
@@ -503,9 +503,28 @@ public sealed override bool Equals(Symbol symbol, TypeCompareKind compareKind)
         {
             if ((object)this == symbol) return true;
 
-            var localFunction = symbol as LocalFunctionSymbol;
-            return (object)localFunction != null
-                && localFunction._syntax == _syntax;
+            return symbol is LocalFunctionSymbol localFunction
+                && localFunction._syntax == _syntax
+                && localFunction.ContainingSymbol.Equals(ContainingSymbol, compareKind);
+        }
+
+        /// <summary>
+        /// Returns a local function symbol with the given containing symbol. The containing symbol passed should
+        /// be identical to the current symbol, with the exception of nullability modifiers. If the passed symbol
+        /// is already the containing symbol of this symbol, the existing symbol will be returned unmodified.
+        /// </summary>
+        internal LocalFunctionSymbol WithContainingSymbol(Symbol containingSymbol)
+        {
+            Debug.Assert(ContainingSymbol.Equals(containingSymbol, TypeCompareKind.AllNullableIgnoreOptions));
+
+            if (ContainingSymbol.Equals(containingSymbol, TypeCompareKind.ConsiderEverything))
+            {
+                return this;
+            }
+            else
+            {
+                return new LocalFunctionSymbol(_binder, containingSymbol, _syntax);
```

Comment:
> return new LocalFunctionSymbol(_binder, containingSymbol, _syntax); [](start = 15, length = 68)

It looks like this is going to force us to do binding of syntax in order to get all aspects of the local function. Attributes, parameter types, constraints, etc. I think we would want to avoid all that work. Perhaps by creating specializedwrapper symbols instead.

Reply:
I'm not entirely sure that's a bad thing here. Otherwise, we have to create a new system to recreate the local function symbol with all parameters substituted for reparented symbols, and the lazy init mechanism that will require.

---
In reply to: [327920751](https://github.com/dotnet/roslyn/pull/38769#discussion_r327920751) [](ancestors = 327920751)

Reply:
As discussed offline, I'm going to revert the LocalFunction changes in this PR entirely, to be addressed in a followup after equality is fixed.

---
In reply to: [329758111](https://github.com/dotnet/roslyn/pull/38769#discussion_r329758111) [](ancestors = 329758111,327920751)